### PR TITLE
feat: add external pricing sync with LiteLLM source

### PR DESCRIPTION
**Pull request by Regis-RCR**
## Summary

OmniRoute currently covers only ~130 models with hardcoded pricing that goes stale between releases. This PR adds a **3-tier pricing resolution system** inspired by Bifrost's sync pattern:

```
User overrides (PATCH /api/pricing)        → highest priority, never touched
    ↓ fallback
Synced external (namespace pricing_synced)  → populated by auto/manual sync
    ↓ fallback
Hardcoded defaults (getDefaultPricing())    → fallback of last resort
```

### What's included

- **`src/lib/pricingSync.ts`** — Sync engine: fetches LiteLLM pricing data, transforms `$/token → $/1M tokens`, maps providers to OmniRoute aliases, stores in separate `pricing_synced` DB namespace
- **`src/app/api/pricing/sync/route.ts`** — REST API: `POST` (trigger sync), `GET` (status), `DELETE` (clear synced data)
- **`src/lib/db/settings.ts`** — `getPricing()` now merges 3 layers (defaults → synced → user overrides)
- **`src/server-init.ts`** — Optional startup sync when `PRICING_SYNC_ENABLED=true`
- **MCP tool `omniroute_sync_pricing`** — Phase 2 tool with `pricing:write` scope, full audit logging
- **23 tests** — 12 Node test runner (transform logic) + 11 Vitest (MCP schema validation)

### Design decisions

- **Opt-in**: `PRICING_SYNC_ENABLED=false` by default — zero behavior change for existing users
- **User overrides are sacred**: stored in separate `pricing_synced` namespace, never overwrites `pricing` namespace
- **Zero new dependencies**: uses native `fetch()` (Node 18+)
- **Graceful degradation**: sync failure logs a warning and continues with cached/default data
- **Idempotent**: re-running sync produces the same result
- **Periodic sync**: configurable interval via `PRICING_SYNC_INTERVAL` (default: 24h)
- **Source validation**: only allowlisted sources accepted (`litellm`), unknown sources rejected with error

### Configuration

```bash
PRICING_SYNC_ENABLED=true           # Enable sync (default: false)
PRICING_SYNC_INTERVAL=86400         # Sync interval in seconds (default: 24h)
PRICING_SYNC_SOURCES=litellm        # Comma-separated sources (default: litellm)
```

### Review feedback addressed (commit 2)

- Added `.catch()` on initial/periodic sync promises (prevent unhandled rejections)
- Wrapped `JSON.parse` in try-catch for corrupted DB data
- Wrapped `response.json()` for invalid LiteLLM responses
- Validated `PRICING_SYNC_INTERVAL` (guard against NaN/0 tight loops)
- Added source allowlist — unknown sources rejected, empty sources won't clear synced data
- Extracted duplicate merge loop into shared iteration
- Added `data`/`warnings` fields to MCP output schema
- Removed unused import, filtered non-string source entries

## Test plan

- [x] `npm run test` — 12 new unit tests pass (transform, provider mapping, edge cases)
- [x] `npm run test:vitest` — 11 new MCP schema tests pass (tool registration, input validation, scopes)
- [x] `npm run lint` — 0 errors (15 pre-existing warnings unchanged)
- [ ] Manual: `PRICING_SYNC_ENABLED=true` → verify sync on startup
- [ ] Manual: `POST /api/pricing/sync` → verify manual trigger
- [ ] Manual: verify user overrides via `PATCH /api/pricing` are preserved after sync

### CODE REVIEW

File: src/lib/db/settings.ts [hunk 1]
```diff
@@ -104,29 +123,36 @@ export async function getPricing() {
...
       for (const [model, pricing] of Object.entries(models)) {
-        if (!mergedPricing[provider][model]) {
-          mergedPricing[provider][model] = pricing;
-        }
+        mergedPricing[provider][model] = mergedPricing[provider][model]
+          ? { ...(mergedPricing[provider][model] || {}), ...toRecord(pricing) }
+          : pricing;
       }
     }
   }
```
[the remark this answers is not in the document]
Comment (Regis-RCR) [reply src/lib/db/settings.ts]: Fixed in 7db280e — extracted the duplicate merge loops into a shared `for (const layer of [syncedPricing, userPricing])` iteration.

File: src/lib/pricingSync.ts [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+  console.log(`[PRICING_SYNC] Starting periodic sync every ${interval / 1000}s`);
+
+  // Initial sync (non-blocking)
+  syncPricingFromSources().then((result) => {
+    if (result.success) {
+      console.log(
+        `[PRICING_SYNC] Initial sync complete: ${result.modelCount} models from ${result.providerCount} providers`
+      );
+    }
+  });
```
[the remark this answers is not in the document]
Comment (Regis-RCR) [reply src/lib/pricingSync.ts]: Fixed in 7db280e — added `.catch()` on the initial sync promise chain.

File: src/lib/pricingSync.ts [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+
+  syncTimer = setInterval(() => {
+    syncPricingFromSources().then((result) => {
+      if (result.success) {
+        console.log(`[PRICING_SYNC] Periodic sync complete: ${result.modelCount} models`);
+      }
+    });
```
[the remark this answers is not in the document]
Comment (Regis-RCR) [reply src/lib/pricingSync.ts]: Fixed in 7db280e — added `.catch()` on the periodic sync promise chain as well.

File: src/lib/pricingSync.ts [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+ * Fetch raw pricing data from LiteLLM GitHub.
+ */
+export async function fetchLiteLLMPricing(): Promise<Record<string, LiteLLMModelInfo>> {
+  const response = await fetch(LITELLM_PRICING_URL, {
+    signal: AbortSignal.timeout(30000),
+  });
+  if (!response.ok) {
+    throw new Error(`LiteLLM fetch failed [${response.status}]: ${response.statusText}`);
+  }
+  return response.json() as Promise<Record<string, LiteLLMModelInfo>>;
```
[the remark this answers is not in the document]
Comment (Regis-RCR) [reply src/lib/pricingSync.ts]: Fixed in 7db280e — `fetchLiteLLMPricing()` now reads `response.text()` first, then wraps `JSON.parse()` in try-catch with a descriptive error message.

File: src/lib/pricingSync.ts [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+  const rows = db
+    .prepare("SELECT key, value FROM key_value WHERE namespace = 'pricing_synced'")
+    .all();
+  const synced: PricingByProvider = {};
+  for (const row of rows) {
+    const record = toRecord(row);
+    const key = typeof record.key === "string" ? record.key : null;
+    const rawValue = typeof record.value === "string" ? record.value : null;
+    if (!key || rawValue === null) continue;
+    synced[key] = JSON.parse(rawValue) as PricingModels;
```
[the remark this answers is not in the document]
Comment (Regis-RCR) [reply src/lib/pricingSync.ts]: Fixed in 7db280e — `getSyncedPricing()` now wraps `JSON.parse(rawValue)` in try-catch, logs a warning for corrupted entries, and skips them.

File: src/lib/pricingSync.ts [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+  sources?: string[];
+  dryRun?: boolean;
+}): Promise<SyncResult> {
+  const sources = opts?.sources || SYNC_SOURCES;
+  const dryRun = opts?.dryRun ?? false;
+
+  try {
+    let aggregated: PricingByProvider = {};
+
+    for (const source of sources) {
```
[the remark this answers is not in the document]
Comment (Regis-RCR) [reply src/lib/pricingSync.ts]: Fixed in 7db280e — added a `SUPPORTED_SOURCES` allowlist. Unknown sources are now rejected with an explicit error. If at least one valid source exists, unknown ones are reported in a `warnings` field.

File: src/lib/pricingSync.ts [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+  providerCount: number;
+  source: string;
+  dryRun: boolean;
+  data?: PricingByProvider;
+  error?: string;
+}
+
+// ─── Configuration ───────────────────────────────────────
+
+const SYNC_INTERVAL_MS = parseInt(process.env.PRICING_SYNC_INTERVAL || "86400", 10) * 1000;
```
[the remark this answers is not in the document]
Comment (Regis-RCR) [reply src/lib/pricingSync.ts]: Fixed in 7db280e — now validates with `Number.isFinite(parsed) && parsed > 0`, falls back to 86400s (24h) on invalid values.

File: src/lib/pricingSync.ts [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+        const raw = await fetchLiteLLMPricing();
+        const transformed = transformToOmniRoute(raw);
+        // Merge into aggregated
+        for (const [provider, models] of Object.entries(transformed)) {
+          if (!aggregated[provider]) aggregated[provider] = {};
+          Object.assign(aggregated[provider], models);
+        }
+      }
+    }
+
```
[the remark this answers is not in the document]
Comment (Regis-RCR) [reply src/lib/pricingSync.ts]: Fixed in 7db280e — `syncPricingFromSources` now validates sources against an allowlist. Returns `success: false` with error if no valid sources provided, preventing accidental data clearing.

File: src/lib/pricingSync.ts [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+    enabled,
+    lastSync: lastSyncTime,
+    lastSyncModelCount,
+    nextSync:
+      syncTimer && lastSyncTime
+        ? new Date(new Date(lastSyncTime).getTime() + SYNC_INTERVAL_MS).toISOString()
+        : null,
+    intervalMs: SYNC_INTERVAL_MS,
+    sources: SYNC_SOURCES,
+  };
```
[the remark this answers is not in the document]
Comment (Regis-RCR) [reply src/lib/pricingSync.ts:383]: Fixed in 7db280e — added `activeSyncIntervalMs` state variable, updated in `startPeriodicSync()`. `getSyncStatus()` now uses the tracked active interval.

File: src/app/api/pricing/sync/route.ts [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+  try {
+    const body = await request.json().catch(() => ({}));
+    const sources = Array.isArray(body.sources) ? body.sources : undefined;
+    const dryRun = body.dryRun === true;
+
+    const { syncPricingFromSources } = await import("@/lib/pricingSync");
+    const result = await syncPricingFromSources({ sources, dryRun });
+
+    return NextResponse.json(result, { status: result.success ? 200 : 502 });
+  } catch (err) {
```
[the remark this answers is not in the document]
Comment (Regis-RCR) [reply src/app/api/pricing/sync/route.ts:23]: Fixed in 7db280e — `sources` array entries are now filtered to strings only (`.filter((s): s is string => typeof s === 'string')`). Downstream `syncPricingFromSources` validates against the allowlist.

File: open-sse/mcp-server/schemas/tools.ts [hunk 11]
```diff
@@ -723,6 +723,40 @@ export const getSessionSnapshotTool: McpToolDefinition<
...
+    .describe("If true, preview sync results without saving to database"),
+});
+
+export const syncPricingOutput = z.object({
+  success: z.boolean(),
+  modelCount: z.number(),
+  providerCount: z.number(),
+  source: z.string(),
+  dryRun: z.boolean(),
+  error: z.string().optional(),
```
[the remark this answers is not in the document]
Comment (Regis-RCR) [reply open-sse/mcp-server/schemas/tools.ts:744]: Fixed in 7db280e — added optional `data` and `warnings` fields to `syncPricingOutput` schema.

File: open-sse/mcp-server/__tests__/pricingSync.test.ts [hunk 12]
```diff
@@ -0,0 +1,60 @@
+import { describe, it, expect } from "vitest";
+import { z } from "zod";
```
[the remark this answers is not in the document]
Comment (Regis-RCR) [reply open-sse/mcp-server/__tests__/pricingSync.test.ts]: Fixed in 7db280e — removed the unused `z` import.
